apollo-federation/src/operation/optimize.rs

```diff
@@ -989,7 +991,13 @@ impl NamedFragments {
         for (_, fragment) in self.iter_mut() {
             Node::make_mut(fragment).selection_set = fragment
                 .selection_set
-                .retain_fragments(&fragments_to_keep)?;
+                .retain_fragments(&fragments_to_keep)?
+                .normalize(
+                    &fragment.selection_set.type_position,
+                    &fragments_to_keep,
+                    &fragment.schema,
+                    NormalizeSelectionOption::NormalizeRecursively,
```

Comment:
Is recursive normalization necessary here?

JS also does it recursively, but I wonder why we even have this option. Neither JS nor Rust QP use `recursive: false` / `NormalizeSingleSelection` anywhere. Should we remove it?

Reply:
Indeed! It is actually pretty weird as in JS code we never actually specify it so it defaults to `undefined` which is then always coalesced to true (using `recursive ?? true` check). I'll open a follow up item to clean it up.

Reply:
Hmm, there is a `normalize` call after `retain_fragments`
```rust
        let reduced_selection_set = selection_set.retain_fragments(self)?;

        // Expanding fragments could create some "inefficiencies" that we wouldn't have if we
        // hadn't re-optimized the fragments to de-optimize it later, so we do a final "normalize"
        // pass to remove those.
        reduced_selection_set.normalize(
            &reduced_selection_set.type_position,
            self,
            &selection_set.schema,
            NormalizeSelectionOption::NormalizeRecursively,
        )
```
I wonder why that didn't handle this.

Reply:
Or, we need to normalize expanded fragment "definitions"?
In `fn reduce()`:
```rust
        let fragments_to_keep = self.clone();
        for (_, fragment) in self.iter_mut() {
            Node::make_mut(fragment).selection_set = fragment
                .selection_set
                .retain_fragments(&fragments_to_keep)?;
             // <= also normalize here?
        }
```

Reply:
yes it was normalizing selection set but not the fragments (hence the extra normalize after retain fragments)